src/pcre2_jit_compile.c

```diff
@@ -14356,23 +14356,25 @@ return PCRE2_ERROR_JIT_BADOPTION;
 
 if ((re->flags & PCRE2_NOJIT) != 0) return 0;
 
-if (executable_allocator_is_working == 0)
+int executable_allocator_is_working_local = executable_allocator_is_working;
+if (executable_allocator_is_working_local == 0)
```

Comment:
The point of this check is checking once that the allocator is working. If it does not, it disables jit compiling. It can be run in multiple threads. Of course the allocator may fail to allocate memory later, but at least it is working, so there is a chance it can do it.

Reply:
wouldn't it be better to then make the assignment of executable_allocator_is_working as part of the else?
```
if (ptr != NULL)
    {
    SLJIT_FREE_EXEC(((sljit_u8*)(ptr)) + SLJIT_EXEC_OFFSET(ptr), NULL);
    executable_allocator_is_working = 1;
    }
  }
else executable_allocator_is_working = -1;
```
as documented around lines 14362 to 14365, 2 different threads should follow the same branch regardless.  maybe changing the wording of the last sentence might needed to make that more clear.

there is always a possible race condition, but any application that gets PCRE2_ERROR_NOMEMORY from this function is expected to disable JIT, since the only way that would be triggered, is because the system is no longer allowing RWX pages (ex: `SELinux enabled`) and the alternative is to segfault.

Reply:
Most likely this will also work correctly:

```c
if (executable_allocator_is_working == 0)
  {
  /* Checks whether the executable allocator is working. This check
     might run multiple times in multi-threaded environments, but the
     result should not be affected by it. */
  void *ptr = SLJIT_MALLOC_EXEC(32, NULL);
  if (ptr != NULL)
    {
    SLJIT_FREE_EXEC(((sljit_u8*)(ptr)) + SLJIT_EXEC_OFFSET(ptr), NULL);
    executable_allocator_is_working_local = 1;
    }
  else executable_allocator_is_working = -1;
  }

if (executable_allocator_is_working < 0)
  return PCRE2_ERROR_NOMEMORY;
```

Which implementation will we use?

Reply:
The first version has the advantage that there is only one place where the value of the `executable_allocator_is_working` variable is set.  
But it's not very important.

Reply:
I prefer the later, but since you are already touching it might as well change it to so it is a more common tri state (-1 = unknown, 0 = false, 1 = true), which will also clean the last check so it will be instead the more reasonable :

```
if (!executable_alllocator_is_working)
  return PCRE2_ERROR_NOMEMORY;
```

Reply:
Do I understand correctly what you are suggesting to write like this?

```c
static int executable_allocator_is_working = -1;

...

if (executable_allocator_is_working == -1)
  {
  /* Checks whether the executable allocator is working. This check
     might run multiple times in multi-threaded environments, but the
     result should not be affected by it. */
  void *ptr = SLJIT_MALLOC_EXEC(32, NULL);
  if (ptr != NULL)
    {
    SLJIT_FREE_EXEC(((sljit_u8*)(ptr)) + SLJIT_EXEC_OFFSET(ptr), NULL);
    executable_allocator_is_working = 1;
    }
  else executable_allocator_is_working = 0;
  }

if (!executable_allocator_is_working)
  return PCRE2_ERROR_NOMEMORY;
```

Reply:
done